### PR TITLE
Exercise 5

### PR DESCRIPTION
**Pull request by RespectableRuessel**
Pull request for exercise 5.
Replaced InMemoryDatasource with a Neo4j database. Now, results of mutations will be persistent. 

* Implemented connection to Neo4j
* Refactored mutations and queries
* Implemented tests for login and signup

[Link to the review](https://github.com/Systems-Development-and-Frameworks/bug-free-disco/pull/11#pullrequestreview-564803713)

### CODE REVIEW

File: backend/src/database/Neo4JDatasource.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            ? { database: NEO4J_DATABASE, defaultAccessMode: session.WRITE }
+            : {};
+        return this.driver.session(options);
+    }
+
+    public async erase(): Promise<void> {
+        const session = this.createSession();
+        await session.writeTransaction(tx => tx.run("MATCH (a) -[r] -> () DELETE a, r"));
+        await session.writeTransaction(tx => tx.run("MATCH (a) DELETE a"));
+        await session.close();
```
Comment (FrauMauz) [inline backend/src/database/Neo4JDatasource.ts:213]: function erase deletes all entries in the db ... good to have :)

File: backend/src/generated/graphqlgen.ts [hunk 2]
```diff
@@ -249,6 +268,23 @@ export namespace UserResolvers {
...
+    | {
+        fragment: string;
+        resolve: (
+          parent: User,
+          args: {},
+          ctx: Context,
+          info: GraphQLResolveInfo
+        ) => string | Promise<string>;
+      };
+
```
Comment (FrauMauz) [inline backend/src/generated/graphqlgen.ts:287]: so many Resolvers :sweat_smile:

File: backend/src/generated/graphqlgen.ts [hunk 3]
```diff
@@ -308,10 +378,6 @@ export namespace MutationResolvers {
 
   export interface PostInput {
     title: string;
-    author: UserInput;
-  }
-  export interface UserInput {
-    name: string;
```
Comment (FrauMauz) [inline backend/src/generated/graphqlgen.ts:314]: you totally deleted the UserInput? why? what was the need/function of this type ?

File: backend/src/generated/graphqlgen.ts [hunk 4]
```diff
@@ -357,6 +433,40 @@ export namespace MutationResolvers {
...
+    | {
+        fragment: string;
+        resolve: (
+          parent: undefined,
+          args: ArgsSignup,
+          ctx: Context,
+          info: GraphQLResolveInfo
+        ) => string | null | Promise<string | null>;
+      };
+
```
Comment (FrauMauz) [inline backend/src/generated/graphqlgen.ts:469]: Signup and Login Resolvers... :wave:

File: backend/src/generated/graphqlgen.ts [hunk 5]
```diff
@@ -391,6 +501,40 @@ export namespace MutationResolvers {
...
+        ) => string | null | Promise<string | null>)
+      | {
+          fragment: string;
+          resolve: (
+            parent: undefined,
+            args: ArgsSignup,
+            ctx: Context,
+            info: GraphQLResolveInfo
+          ) => string | null | Promise<string | null>;
+        };
```
Comment (FrauMauz) [inline backend/src/generated/graphqlgen.ts:537]: definition of the parameters of the new Types. yep that needed

File: backend/src/generated/tmp-resolvers/Mutation.ts [hunk 6]
```diff
@@ -10,5 +10,11 @@ export const Mutation: MutationResolvers.Type = {
   },
   upvote: (parent, args, ctx) => {
     throw new Error("Resolver not implemented");
+  },
+  login: (parent, args, ctx) => {
+    throw new Error("Resolver not implemented");
+  },
+  signup: (parent, args, ctx) => {
+    throw new Error("Resolver not implemented");
```
Comment (FrauMauz) [inline backend/src/generated/tmp-resolvers/Mutation.ts:18]: ... and of cause the mutations for signing up and login

File: backend/src/index.ts [hunk 7]
```diff
@@ -1,7 +1,9 @@
-import { createServer } from "./server";
+import {closeConnection, createServer} from "./server";
```
Comment (FrauMauz) [inline backend/src/index.ts:1]: alright

File: backend/src/interfaces/Post.ts [hunk 8]
```diff
@@ -4,11 +4,11 @@ export interface Post {
     id: string;
     title: string;
     votes: number;
-    author: User;
-    lastVoted: Votes[];
+    author: string;
```
Comment (FrauMauz) [inline backend/src/interfaces/Post.ts:7]: ahh you changed the type from typte user to type string. seems to be easier right?

File: backend/src/interfaces/Post.ts [hunk 9]
```diff
@@ -4,11 +4,11 @@ export interface Post {
...
+    lastVoted?: Votes[];
 }
 
 interface Votes {
-    username: string;
```
Comment (FrauMauz) [inline backend/src/interfaces/Post.ts:12]: you don't track the users that did the votes anymore ... also good choice

File: backend/src/interfaces/User.ts [hunk 10]
```diff
@@ -1,6 +1,9 @@
 import { Post } from "./Post";
 
 export interface User {
+    id: string;
     name: string;
+    email: string;
+    password: string;
```
Comment (FrauMauz) [inline backend/src/interfaces/User.ts:7]: alright.! needed for the login...

File: backend/src/resolvers/nested/post.ts [hunk 11]
```diff
@@ -2,8 +2,10 @@ import { PostResolvers } from '../../generated/graphqlgen';
...
 export const Post: PostResolvers.Type = {
     ...PostResolvers.defaultResolvers,
-    // @ts-ignore
-    author: (parent, _args, { dataSources }) => {
-        return dataSources.databaseAPI.getAuthor(parent.id);
+    author: (parent, _args, ctx) => {
+        return ctx.dataSources.databaseAPI.getAuthor(parent.id);
     },
+    votes: (parent, args, ctx) => {
+        return ctx.dataSources.databaseAPI.countVotesOfOnePost(parent.id, true);
```
Comment (FrauMauz) [inline backend/src/resolvers/nested/post.ts:9]: countVotesOfOnePost where did you do this before ?

File: backend/src/resolvers/user.resolver.ts [hunk 12]
```diff
@@ -8,3 +10,26 @@ export const UserQueryResolver: Pick<QueryResolvers.Type
...
+export const UserMutationResolver: Pick<MutationResolvers.Type
+    , 'login'
+    | 'signup'
+    > = {
+    // @ts-ignore
+    login: async (parent, {email, password}, {dataSources}) => {
+        email = email.toLocaleLowerCase();
+        const user = await dataSources.databaseAPI.getUserPerEmail(email);
+        if (!user) return null;
+        if (user.password !== generateSHA512Hash(password)) return null;
```
Comment (FrauMauz) [inline backend/src/resolvers/user.resolver.ts:23]: ahhh here you use the new functions for hashing and check them (:star2:)

File: backend/src/schema.graphql [hunk 13]
```diff
@@ -6,7 +6,9 @@ type Post {
 }
 
 type User {
-    name: ID!
+    id: ID!,
+    name: String!,
+    email: String!,
```
Comment (FrauMauz) [inline backend/src/schema.graphql:11]: updated schema... everything ok

File: backend/src/test/post.spec.ts [hunk 14]
```diff
@@ -1,44 +1,26 @@
...
+import {Neo4JDatasource} from "../database/Neo4JDatasource";
+import {auth, driver} from "neo4j-driver";
+import {NEO4J_PASSWORD, NEO4J_URI, NEO4J_USER} from "../config/env.config";
+
+const drv = driver(NEO4J_URI, auth.basic(NEO4J_USER, NEO4J_PASSWORD));
+const db: Neo4JDatasource = new Neo4JDatasource(drv);
 const server = createServer(db);
 const {query, mutate} = createTestClient(server);
 
-const getDummyUsers = () => [
```
Comment (FrauMauz) [inline backend/src/test/post.spec.ts:10]: detailed because not needed anymore

File: backend/src/test/post.spec.ts [hunk 15]
```diff
@@ -1,44 +1,26 @@
...
+import { generateSHA512Hash } from '../authorization/cryptography';
+import { Post } from "../interfaces/Post";
+import { User } from "../interfaces/User";
+import { verifyToken } from "../authorization/authorization";
+import {Neo4JDatasource} from "../database/Neo4JDatasource";
+import {auth, driver} from "neo4j-driver";
+import {NEO4J_PASSWORD, NEO4J_URI, NEO4J_USER} from "../config/env.config";
+
+const drv = driver(NEO4J_URI, auth.basic(NEO4J_USER, NEO4J_PASSWORD));
+const db: Neo4JDatasource = new Neo4JDatasource(drv);
```
Comment (FrauMauz) [inline backend/src/test/post.spec.ts:14]: new Neo4JDatasource(drv); !!! good

File: backend/src/test/post.spec.ts [hunk 16]
```diff
@@ -87,16 +69,27 @@ describe('queries', () => {
...
 describe('mutations', () => {
+    beforeEach(async () => {
+        const SIGNUP_USER = gql`
+            mutation ($name: String!, $email: String!, $password: String!) {
+                signup(name: $name, email: $email, password: $password)
+            }
+        `;
+
+        const data = await mutate({ mutation: SIGNUP_USER, variables: {
+                name:"Guy", email:"GUY@GUY.de", password:"123456789"
```
Comment (FrauMauz) [inline backend/src/test/post.spec.ts:80]: for testing.. ok

File: backend/test.env [hunk 17]
```diff
@@ -0,0 +1,5 @@
+TOKEN_SECRET=TEST123456
+NEO4J_URI=bolt://18.204.8.8:33335
+NEO4J_USER=neo4j
+NEO4J_PASSWORD=twos-pages-diagnosis
+NEO4J_DATABASE=
```
Comment (FrauMauz) [inline backend/test.env:5]: test Variables

Comment (FrauMauz) [reply backend/test.env:5]: ⭐ For requesting a review and reviewing another team's PR.
You asked review from us.. and we are sure that you will review us as soon as you have our request. so this star will be yours for sure...

File: backend/src/types/context.ts [hunk 18]
```diff
@@ -1,2 +1,8 @@
+import {DatasourceAPI} from "../database/DatasourceAPI";
+
 export interface Context {
+    userId: string;
+    dataSources: {
+        databaseAPI: DatasourceAPI
+    };
 }
```
Comment (FrauMauz) [inline backend/src/types/context.ts:8]: Interface for db ... good

File: backend/src/server.ts [hunk 19]
```diff
@@ -2,47 +2,76 @@ import {ApolloServer, gql} from "apollo-server";
...
+        name: 'Daniel',
+        email: 'test2@test.de',
+        password: generateSHA512Hash('test123')
     },
     {
-        name: 'Robin'
+        id: '789',
+        name: 'Robin',
+        email: 'test3@test.de',
+        password: generateSHA512Hash('test123')
```
Comment (FrauMauz) [inline backend/src/server.ts:39]: some test users with their data and hash passwords...

**Review by FrauMauz — APPROVED** [review]
schon sehr komplex mittlerweile... man sieht, dass viel Arbeit drin steckt !

File: backend/.gitignore [hunk 20]
```diff
@@ -1,3 +1,6 @@
 .idea
 node_modules
 lib
+.env
+
+dev.env
```
[the remark this answers is not in the document]
Comment (SarahLM) [reply backend/.gitignore]: instead of ignore the .env , you could use git-crypt

**Review by medizinmensch — APPROVED** [review]
Hi Team!

I really like your thorough Readme, this made it much easier for me to review it! Thanks!

It looks like you have flaky tests. If I do `npm run test`, I get different errors every time - sometimes also no error.
This is because you're running your test files in parallel - jest is doing that by default. I recommend adding a `--runInBand` [here](https://github.com/Systems-Development-and-Frameworks/no-js-pls/blob/0f973ab97b6e39c955ad53c3612dd2e5d1fdc197/backend/package.json#L10) to change that. 
```
"test": "dotenv -e test.env -- jest --runInBand"
```
Otherwise a change in your database from one file affects the other. 
Once that was solved it looked quite smoothly. 

One thing I wonder though... Are you the group that didn't use a locally hosted db at all? So did you only use a hosted/sandbox one online? Is it possible that your `test.env` contains actual connection data? 😅 
As you can guess, this is not how it's supposed to be. 
I can only emphasize that starting one with docker is quite easy once you got docker installed. For instance with:
```
docker run -p 7474:7474 -p 7687:7687 --env=NEO4J_AUTH=NEO4J_USERNAME/NEO4J_PASSWORD neo4j:latest
```

:x: So feel free to clarify, but it looks like I can't give you a point here:
> :star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.

:x: Secondly, I do not see that you deny your requests to your subschema. It would be easiest to set a default deny rule in your main schema so that everything not in that schema gets denied.
> :star: You deny the client to   call mutations and queries of your subschema directly.

To conclude:

:star: For choosing a scenario   and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.
:star: :star: Your software   tests are free of side effects.
:star:   Your database is never left in an invalid state.
:star:   For requesting a review and reviewing another team's PR.

That makes 8/10 :star:

Gif

**Comment by RespectableRuessel** [comment]
> 
> 
> Hi Team!
> 
> I really like your thorough Readme, this made it much easier for me to review it! Thanks!
> 
> It looks like you have flaky tests. If I do `npm run test`, I get different errors every time - sometimes also no error.
> This is because you're running your test files in parallel - jest is doing that by default. I recommend adding a `--runInBand` [here](https://github.com/Systems-Development-and-Frameworks/no-js-pls/blob/0f973ab97b6e39c955ad53c3612dd2e5d1fdc197/backend/package.json#L10) to change that.
> 
> ```
> "test": "dotenv -e test.env -- jest --runInBand"
> ```
> 
> Otherwise a change in your database from one file affects the other.
> Once that was solved it looked quite smoothly.
> 
> One thing I wonder though... Are you the group that didn't use a locally hosted db at all? So did you only use a hosted/sandbox one online? Is it possible that your `test.env` contains actual connection data? 😅
> As you can guess, this is not how it's supposed to be.
> I can only emphasize that starting one with docker is quite easy once you got docker installed. For instance with:
> 
> ```
> docker run -p 7474:7474 -p 7687:7687 --env=NEO4J_AUTH=NEO4J_USERNAME/NEO4J_PASSWORD neo4j:latest
> ```
> 
> ❌ So feel free to clarify, but it looks like I can't give you a point here:

The problem is that it seems that Neo4j can not be tested locally on your normal machine. I've written a Moodle Post about it because I looked up the issue and nobody who worked with Neo4j had a actual answer. Robert suggested then to mock out the session but then you're miss testing the connection and exception handling when your database fails which is not desirable.

The only option we found (without docker) was the sandbox that was presented in the lecture. The problem with docker is that I (for example) cannot use docker because it requires "Hyper-V and Containers Windows features must be enabled" . This is not possible for me because I am using a different hypervisor which is not compatible with Hyper-V.

Unfortunately Neo4j's architecture design is bad for using multiple instances because it runs an own service that (at least in free mode) cannot automatically setup a second database for testing.
There is a in-memory API but it is only accessible through the Java library. Also running a local instance is hard because the god-damn paths include a GUID which is different between users that leads to the problem that a configuration file will not work with other users and on the CI because the paths to the Neo4j file differ.

This was the original post:
> Überall steht, dass das leider nicht "einfach" möglich ist [1]. Für die Java / C# Anbindung gibt es die ImpermanentGraphDatabase [2, 3], die es aber nicht für JS oder andere Sprachen implementiert ist.

> Empfohlen wird das "Mocken" der Aufrufe, welches uns aber wieder nur zur InMemoryDatasource führt, die wir ja nicht nutzen sollen. Die paar "Lösungen" die es gibt scheinen zusätzlich eine Bearbeitung der neo4j config zu benötigen [4], welches auf der CI kompliziert wird, da die lokalen neo4j Pfade nicht übereinstimmen.


> [1] https://stackoverflow.com/questions/49348839/how-to-unit-test-neo4j-in-python

> [2] https://github.com/neo4j/neo4j-javascript-driver/issues/211

> [3] https://neo4j.com/docs/java-reference/current/java-embedded/include-neo4j/

> [4] https://stackoverflow.com/questions/45784232/how-to-create-new-database-in-neo4j

**Pull request merged**